### PR TITLE
docs: Workflow + Skopeo + Approval + Image Tag at CI + Image Tag + Image Digest at CD

### PR DESCRIPTION
**Pull request by ashokdevtron**
**Completed**:
- Workflow updated with snapshots
- CI document restructured and updated (Image Tag included)
- CD document restructured and updated (Tag and Approval included)
- Copy Container Image plugin added (Skopeo)
- RBAC
- Image Digest at CD included and GC added

### CODE REVIEW

**Review by abhibhaw — CHANGES_REQUESTED** [review]
Better to include latest images having digest options as well. And maybe we can handle that too with this. Also we need to give some ideal pipelines (Devtron recommended) with this doc

File: docs/user-guide/creating-application/workflow/cd-pipeline.md [hunk 1]
```diff
@@ -287,6 +319,7 @@ Please follow the steps mentioned below to create sequential pipelines:
 2. To add another CD Pipeline sequentially after previous one, again click on + sign on the last CD pipeline.
 3. Similarly, you can add multiple CD pipelines by clicking + sign of the last CD pipeline, each deploying in different environments.
 
-![](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/creating-application/workflow-cd-pipeline/sequential-workflow.jpg)
+![Figure 13: Adding Multiple CD Pipelines](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/creating-application/workflow-cd-pipeline/sequential-workflow.jpg)
```
Comment (uxarya-d) [inline docs/user-guide/creating-application/workflow/cd-pipeline.md:322]: Would be great if the image could depict sequential cd pipelines to staging --> production.

File: docs/user-guide/creating-application/workflow/ci-pipeline.md [hunk 2]
```diff
@@ -164,16 +137,65 @@ Select the appropriate filter and pass the matching condition as a regular expre
...
-> You can also provide some stages on the Devtron tool’s console and some stages in the devtron-ci.yaml file. But stages defined through the `Devtron` dashboard are
-> first executed then the stages defined in the `devtron-ci.yaml` file.
->
-> **(b)** The total timeout for the execution of the CI pipeline is by default set as 3600 seconds. This default timeout is configurable according to the use case. The timeout can be edited in the configmap of the orchestrator service in the env variable as `env:"DEFAULT_TIMEOUT" envDefault:"3600"`
+{% hint style="info" %}
+**(a)** You can provide pre-build and post-build stages via the Devtron tool’s console or can also provide these details by creating a file `devtron-ci.yaml` inside your repository. There is a pre-defined format to write this file. And we will run these stages using this YAML file. You can also provide some stages on the Devtron tool’s console and some stages in the devtron-ci.yaml file. But stages defined through the `Devtron` dashboard are first executed then the stages defined in the `devtron-ci.yaml` file.
```
Comment (uxarya-d) [inline docs/user-guide/creating-application/workflow/ci-pipeline.md:141]: There is a mention about `devtron-ci.yaml` file which has a predefined format, but no mention of where can I find it or how can I use it.

File: docs/user-guide/creating-application/workflow/ci-pipeline.md [hunk 3]
```diff
@@ -164,16 +137,65 @@ Select the appropriate filter and pass the matching condition as a regular expre
...
+
+**(b)** The total timeout for the execution of the CI pipeline is by default set as 3600 seconds. This default timeout is configurable according to the use case. The timeout can be edited in the configmap of the orchestrator service in the env variable as `env:"DEFAULT_TIMEOUT" envDefault:"3600"`
+{% endhint %}
+
+##### Scan for Vulnerabilities
```
Comment (uxarya-d) [inline docs/user-guide/creating-application/workflow/ci-pipeline.md:146]: The scan option is only available if either of the Vulnerability Scanning integrations (Clair/Trivy) are installed via the Devtron Stack Manager section in the dashboard.

File: docs/user-guide/creating-application/workflow/ci-pipeline.md [hunk 4]
```diff
@@ -164,16 +137,65 @@ Select the appropriate filter and pass the matching condition as a regular expre
...
+
+To perform the security scan after the container image is built, enable the **Scan for vulnerabilities** toggle in the build stage.
+
+![](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/creating-application/workflow-ci-pipeline/scan-for-vulnerabilities-v2.png)
```
Comment (uxarya-d) [inline docs/user-guide/creating-application/workflow/ci-pipeline.md:150]: Image link seems to be broken

File: docs/user-guide/creating-application/workflow/ci-pipeline.md [hunk 5]
```diff
@@ -164,16 +137,65 @@ Select the appropriate filter and pass the matching condition as a regular expre
...
+
+    ![](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/creating-application/workflow-ci-pipeline/build-history.jpg)
+
+    * **Docker Registry**
+
+    ![](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/creating-application/workflow-ci-pipeline/docker-image.jpg)
+
+    * **CD Pipeline (Image Selection)**
+
+    ![](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/creating-application/workflow-ci-pipeline/build-history.jpg)
```
Comment (uxarya-d) [inline docs/user-guide/creating-application/workflow/ci-pipeline.md:190]: I suppose this screenshot would be updated to the select image screenshot.

File: docs/user-guide/creating-application/workflow/plugins/copy-container-image.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
+## Copy Container Image
+
+This plugin copies a container image from the source to destination. The pushing of image can be between repositories of the same container registry or between repositories of different container registry.
```
Comment (uxarya-d) [inline docs/user-guide/creating-application/workflow/plugins/copy-container-image.md:3]: Suggestion: This plugin copies a container image to a desired container repository.

File: docs/user-guide/creating-application/workflow/plugins/copy-container-image.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
+## Copy Container Image
```
Comment (uxarya-d) [inline docs/user-guide/creating-application/workflow/plugins/copy-container-image.md:1]: Can we mention a use case when using this plugin would be relevant.

File: docs/user-guide/creating-application/workflow/plugins/copy-container-image.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+    ![](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/creating-application/workflow-cd-pipeline/image-destination.jpg)
+
+    * **registry-name** is the name you gave to your container registry while adding it in [Global Configuration → OCI/Container Registry](../../../global-configurations/container-registries.md#add-container-registry).
+
+    * **user-name** is the your account name registered with you container registry, e.g., DockerHub.
+
+    * **repository-name** is the name of the respository within your container registry that hosts the container images of your application.
+
+6. Click **Update Pipeline**.
```
Comment (uxarya-d) [inline docs/user-guide/creating-application/workflow/plugins/copy-container-image.md:33]: Can we mention & hyperlink the Custom image tag pattern option in CD Pipeline here (which we have already documented)? Would help the reader connect the dots.

File: docs/user-guide/deploying-application/triggering-cd.md [hunk 9]
```diff
@@ -1,70 +1,105 @@
...
+
+1. Navigate to the `Build & Deploy` page, and click the **Approval for deployment** icon.
+
+    ![Figure 4: Approval Button](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/deploying-application/triggering-cd/deployment-approval-button-v2.jpg)
+
+2. Click the **Request Approval** button present on the image for which you want to request an approval and click **Submit Request**.
+
+    ![Figure 5: Requesting Approval](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/deploying-application/triggering-cd/request-approval-v2.jpg)
+
+    In case you have configured [SES or SMTP on Devtron](../global-configurations/manage-notification.md#notification-configurations), you can directly choose the approver(s) from the list of approvers as shown below.
```
Comment (uxarya-d) [inline docs/user-guide/deploying-application/triggering-cd.md:39]: Please mention that the users selected here will be sent an approval request via email. Although, any user with approver permission for app + env in question would be able to approve the image.

File: docs/user-guide/deploying-application/triggering-cd.md [hunk 10]
```diff
@@ -1,70 +1,105 @@
...
+2. You will find all the approved images listed under the `Approved images` section. From the list, you can select the desired image and deploy it to your environment.
 
-After the images have been approved, navigate to the `Build and Deploy` tab and click on `Select Image`. You will find all the approved images listed under the `Approved Images` section. From this section, you can select the desired approved image and deploy it to the environment.
+    ![Figure 15: List of Approved Images](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/deploying-application/triggering-cd/approved-images-v2.jpg)
 
-![](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/deploying-application/triggering-cd/approved-images.jpg)
+3. You can view the status of current deployment in the `App Details` tab. 
 
-The status of the current deployment can be viewed by clicking **App Details** that will show the `Progressing`state for 1-2 minutes and then gradually shows `Healthy` state, based on the deployment strategy.
+    ![Figure 16: 'App Details' Screen](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/deploying-application/triggering-cd/app-status.jpg)
```
Comment (uxarya-d) [inline docs/user-guide/deploying-application/triggering-cd.md:98]: This is a really old screenshot, I suppose we'll update it. :)

**Review by uxarya-d — CHANGES_REQUESTED** [review]
Added review comments

File: docs/user-guide/creating-application/workflow/cd-pipeline.md [hunk 11]
```diff
@@ -3,208 +3,240 @@ Once you are done creating your CI pipeline, you can start building your CD pipe
...
+### 2. Deployment Strategy
 
-Click on `Create Pipeline` to create a CD pipeline. 
-
-> One can have a single CD pipeline or multiple CD pipelines connected to the same CI Pipeline. Each CD pipeline corresponds to only one environment, or in other words, any single environment of an application can have only one CD pipeline.
-So, the images created by the CI pipeline can be deployed into multiple environments through different CD pipelines originating from a single CI pipeline.
-If you already have one CD pipeline and want to add more, you can add them by clicking on the `+` sign and then choosing the environment in which you want to deploy your application. Once a new CD Pipeline is created for the environment of your choosing, you can move ahead and configure the CD pipeline as required.
-Your CD pipeline can be configured for the pre-deployment stage, the deployment stage, and the post-deployment stage. You can also select the deployment strategy of your choice. You can add your configurations as explained below:
+Devtron supports 4 types of deployment strategies. Click on `Add Deployment strategy` and select from the available options:
```
[the remark this answers is not in the document]
Comment (ashokdevtron) [reply docs/user-guide/creating-application/workflow/cd-pipeline.md]: Created a table

File: docs/user-guide/creating-application/workflow/cd-pipeline.md [hunk 12]
```diff
@@ -263,9 +295,9 @@ canary:
 | `maxSurge` | It defines the maximum number of replicas the rollout can create to move to the correct ratio set by the last setWeight |
 | `maxUnavailable` | The maximum number of pods that can be unavailable during the update |
 | `setWeight` | It is the required percent of pods to move to the next step |
-| `duration` | It is used to set the duration to wait to move to the next step. |
+| `duration` | It is used to set the duration to wait to move to the next step |
 
-**Recreate**
+#### 4. Recreate
 
 The recreate strategy is a dummy deployment that consists of shutting down version A then deploying version B after version A is turned off. A recreate deployment incurs downtime because, for a brief period, no instances of your application are running. However, your old code and new code do not run at the same time.
```
[the remark this answers is not in the document]
Comment (ashokdevtron) [reply docs/user-guide/creating-application/workflow/cd-pipeline.md:320]: Checked with DevOps. In 'Recreate' strategy, there is nothing present for a user to configure.

File: docs/user-guide/creating-application/workflow/cd-pipeline.md [hunk 13]
```diff
@@ -3,208 +3,240 @@ Once you are done creating your CI pipeline, you can start building your CD pipe
...
 
-## Deployment Strategies
+## Extras
 
-A deployment strategy is a way to make changes to an application, without downtime in a way that the user barely notices the changes. There are different types of deployment strategies like Blue/green Strategy, Rolling Strategy, Canary Strategy, Recreate Strategy. These deployment configuration-based strategies are discussed in this section.
+### Deployment Strategies
 
-**Blue Green Strategy**
+A deployment strategy is a way to make changes to an application, without downtime, in a way that a user would barely notice the changes. Several deployment configuration-based strategies are discussed in this section.
+
```
[the remark this answers is not in the document]
Comment (ashokdevtron) [reply docs/user-guide/creating-application/workflow/cd-pipeline.md:253]: Added a table in fig 2

File: docs/user-guide/creating-application/workflow/cd-pipeline.md [hunk 14]
```diff
@@ -3,208 +3,240 @@ Once you are done creating your CI pipeline, you can start building your CD pipe
...
+Post-deployment stages are similar to pre-deployment stages. The difference is, pre-deployment executes before the CD pipeline execution and post-deployment executes after the CD pipeline execution. The configuration of post-deployment stages is similar to the pre-deployment stages.
 
-Once you have configured the CD pipeline, click on `Create Pipeline` to save it. You can see your newly created CD Pipeline on the Workflow tab attached to the corresponding CI Pipeline.
+You can use ConfigMap and Secrets in post deployments as well, as defined in the Pre-Deployment stages.
 
+![Figure 6: Post-deployment Stage](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/creating-application/workflow-cd-pipeline/cd_post_build.jpg)
 
-### 7. Execute in Application Environment
+
+##### Execute in Application Environment
```
[the remark this answers is not in the document]
Comment (ashokdevtron) [reply docs/user-guide/creating-application/workflow/cd-pipeline.md]: Now this is moved to **Extras** section. Tried clubbing it with Pre-CD, but looked very cluttered.

File: docs/user-guide/creating-application/workflow/cd-pipeline.md (hunk 1 above)
Comment (ashokdevtron) [reply docs/user-guide/creating-application/workflow/cd-pipeline.md]: Will fix this in upcoming commits

File: docs/user-guide/global-configurations/pull-image-digest.md [hunk 15]
```diff
@@ -0,0 +1,52 @@
+# Pull Image Digest
+
+## Introduction
+
+Devtron offers the option to [pull container images using digest](../creating-application/workflow/cd-pipeline.md#pull-container-image-with-image-digest). Though it can be enabled by an application-admin for a given CD Pipeline, Devtron also allows super-admins to enable pull image digest at environment level.
+
+This helps in better governance and less repetitiveness if you wish to manage pull image digest for multiple applications across environments.
```
Comment (abhibhaw) [inline docs/user-guide/global-configurations/pull-image-digest.md:7]: What this feature actually does? What are the drawbacks? Why should I enable this?

**Review by abhibhaw — CHANGES_REQUESTED** [review]
Overall the content is great, and we have covered all the topics in detail, one concern I have is repetition. A lot of same / similar content in repeated this will make it harder for us to manage and for users it may be confusing as while I was reading every time it was in my mind I think I read this few lines above.

File: docs/user-guide/creating-application/workflow/cd-pipeline.md [hunk 16]
```diff
@@ -1,210 +1,264 @@
...
+If you want to run it within your application, tick the `Execute tasks in application environment` checkbox. Otherwise, leave it unchecked to run it within the Devtron build cluster. By default, this option is disabled, so refer [Executing in Application Environment](#enabling-execution-in-application-environment) to know the process of enabling it.
 
-Once you are done defining Config Maps & Secrets, you will get them as a drop-down in the pre-deployment stage and you can select them as part of your pre-deployment stage.
+#### Deployment Stage
 
-These `Pre-deployment CD / Post-deployment CD` pods can be created in your deployment cluster or the devtron build cluster. It is recommended that you run these pods in the Deployment cluster so that your scripts \(if there are any\) can interact with the cluster services that may not be publicly exposed.
+##### Pipeline Name
 
-If you want to run it inside your application, then you have to check the `Execute in application Environment` option else leave it unchecked to run it within the Devtron build cluster.
+Pipeline name will be auto-generated; however, you are free to modify the name as per your requirement.
```
[the remark this answers is not in the document]
Comment (ashokdevtron) [reply docs/user-guide/creating-application/workflow/cd-pipeline.md:140]: This one is not repeated.

File: docs/user-guide/creating-application/workflow/cd-pipeline.md [hunk 17]
```diff
@@ -1,210 +1,264 @@
...
 3. Select the number of approvals required for each deployment.
 
-### 5. Deployment stage
+To know more about the approval process, refer [Triggering CD](../../deploying-application/triggering-cd.md#manual-approval-for-deployment). 
 
-**\(a\) Deploy to Environment**
+##### Custom Image Tag Pattern
 
-Select the environment where you want to deploy your application. Once you select the environment, it will display the `Namespace` corresponding to your selected environment automatically.
+This feature helps you append custom tags (e.g., `v1.0.0`) to readily distinguish container images within your repository.
```
[the remark this answers is not in the document]
Comment (ashokdevtron) [reply docs/user-guide/creating-application/workflow/cd-pipeline.md]: replaced with 'applied'

File: docs/user-guide/creating-application/workflow/cd-pipeline.md [hunk 18]
```diff
@@ -1,210 +1,264 @@
...
+{% hint style="warning" %}
+### Who Can Perform This Action?
+Users need to have Admin permission or above (along with access to the environment and application) to enable this option. However, this option will be non-editable in case the super-admin has enabled [pull image digest in Global Configurations](../../global-configurations/pull-image-digest.md).
+{% endhint %}
 
-To enable the "Execute in application environment" option, follow these steps:
+#### Post-Deployment Stage
 
-1. Go to the chart store and search for the devtron-in-clustercd chart.
+If you need to run any actions for e.g., run actions like closure of Jira ticket or provide secrets after the deployment, you can configure such actions in the post-deployment stages.
```
[the remark this answers is not in the document]
Comment (ashokdevtron) [reply docs/user-guide/creating-application/workflow/cd-pipeline.md]: Old content by unknown author. Replaced with your suggestion.

File: docs/user-guide/creating-application/workflow/ci-pipeline.md [hunk 19]
```diff
@@ -1,62 +1,58 @@
...
+For Devtron version older than v0.4.0, please refer the [CI Pipeline (legacy)](./ci-pipeline-legacy.md) page.
+{% endhint %}
+
+## Creating CI Pipeline
 
 A CI Workflow can be created in one of the following ways:
 
 * [Build and Deploy from Source Code](#1.-build-and-deploy-from-source-code)
 * [Linked Build Pipeline](#2.-linked-build-pipeline)
 * [Deploy Image from External Service](#3.-deploy-image-from-external-service)
```
[the remark this answers is not in the document]
Comment (ashokdevtron) [reply docs/user-guide/creating-application/workflow/ci-pipeline.md:13]: Mentioned the two for now in bullets. But its detailed explanation will be picked as another task and not in this PR. Added it in my ToDo list.

image

File: docs/user-guide/global-configurations/pull-image-digest.md (hunk 15 above)
Comment (ashokdevtron) [reply docs/user-guide/global-configurations/pull-image-digest.md:9]: Correct. All of these were already explained in the CD Pipeline doc. The hyperlink was given, but now the line has been replaced and made more prominent as follows: _Refer [CD Pipeline - Image Digest](../creating-application/workflow/cd-pipeline.md#pull-container-image-with-image-digest) to know the purpose it serves._

File: docs/user-guide/creating-application/workflow/ci-pipeline.md [hunk 20]
```diff
@@ -138,7 +140,7 @@ Select the appropriate filter and pass the matching condition as a regular expre
 Select **Create Pipeline**.
 
 {% hint style="info" %}
-The total timeout for the execution of the CI pipeline is by default set as 3600 seconds. This default timeout is configurable according to the use case. The timeout can be edited in the configmap of the orchestrator service in the env variable as `env:"DEFAULT_TIMEOUT" envDefault:"3600"`
+The total timeout for the execution of the CI pipeline is by default set as 3600 seconds. This default timeout is configurable according to the use case (refer [Build Infra](https://docs.devtron.ai/global-configurations/build-infra)). The timeout can be edited in the configmap of the orchestrator service in the env variable as `env:"DEFAULT_TIMEOUT" envDefault:"3600"`
```
Comment (abhibhaw) [inline docs/user-guide/creating-application/workflow/ci-pipeline.md:143]: i guess after build infra story, this env variable has been removed. please confirm @gireesh-devtron

**Review by abhibhaw — CHANGES_REQUESTED** [review]

File: docs/user-guide/creating-application/workflow/ci-pipeline.md (hunk 20 above)
Comment (uxarya-d) [reply docs/user-guide/creating-application/workflow/ci-pipeline.md:143]: From Gireesh: "that variable is not removed but timeout for builds now is no longer derived from that variable.
currently that variable is being used for job pods."

**Review by uxarya-d — APPROVED** [review]

File: docs/user-guide/creating-application/workflow/ci-pipeline.md (hunk 20 above)
Comment (ashokdevtron) [reply docs/user-guide/creating-application/workflow/ci-pipeline.md]: Done

**Pull request merged**